{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 1:Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n",
      "NO AGI Locus Code\n"
     ]
    }
   ],
   "source": [
    "Germplasm= open(\"Germplasm.tsv\", \"r\")\n",
    "Germplasm.seek(0)\n",
    "#print(Germplasm.readlines())\n",
    "for line in Germplasm.readlines():\n",
    "    matchObj=re.search(r'\\n([AT].[G].+\\t)',line)\n",
    "    if matchObj:\n",
    "        Locus=matchObj.group()\n",
    "    else:\n",
    "        print(\"NO AGI Locus Code\")\n",
    "Germplasm.close()\n",
    "        \n",
    "LocusGene=open(\"LocusGene.tsv\",\"r\")\n",
    "LocusGene.seek(0)\n",
    "#print(LocusGene.readlines())\n",
    "for line in LocusGene.readlines():\n",
    "    matchObj2=re.search(r'\\n([AT].[G].+\\t)',line)\n",
    "    if matchObj2:\n",
    "        Locus2=matchObj2.group()\n",
    "    else:\n",
    "        print(\"NO AGI Locus Code\")\n",
    "LocusGene.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 2: \n",
    "\n",
    " It should have two tables - one for each of the two data files.\n",
    " The two tables should be linked in a 1:1 relationship\n",
    " you may use either sqlMagic or pymysql to build the database\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "(pymysql.err.ProgrammingError) (1007, \"Can't create database 'Exam_2'; database exists\")\n",
      "[SQL: CREATE database Exam_2]\n",
      "(Background on this error at: http://sqlalche.me/e/f405)\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus</td>\n",
       "        <td>varchar(10)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Gene</td>\n",
       "        <td>varchar(10)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ProteinLength</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('Locus', 'varchar(10)', 'NO', '', None, ''),\n",
       " ('Gene', 'varchar(10)', 'NO', '', None, ''),\n",
       " ('ProteinLength', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases\n",
    "%sql CREATE database Exam_2\n",
    "%sql use Exam_2;\n",
    "%sql show tables\n",
    "%sql CREATE TABLE germplasm_table(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, germplasm VARCHAR(100) NOT NULL, phenotype VARCHAR(1000) NOT NULL, pumbed INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm_table\n",
    "%sql show tables\n",
    "%sql CREATE TABLE LocusGene_table (id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, Gene VARCHAR(10) NOT NULL, ProteinLength INTEGER NOT NULL);\n",
    "%sql describe LocusGene_table\n",
    "%sql show tables\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Problem 3: \n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT1G01040\",\"CS3828\", \"Increased abundance of miRNA precursors.\",\"17369351\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT1G01060\",\"lhy-101\",\"The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\",\"16891401\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT1G01140\",\"SALK_058629\",\"hypersensitive to low potassium media\",\"17486125\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT1G01220\",\"SALK_012400C\",\"fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\",\"18199744\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT2G03720\",\"SALK_042433\",\"Multiple straight hairs\",\"16367956\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT2G03800\",\"gek1-1\",\"Ethanol hypersensitivity.\",\"15215505\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT2G04240\",\"xerico\",\"Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\",\"17933900\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT2G05210\",\"pot1-1\",\"No visible phenotype.\",\"17627276\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02130\",\"rpk2-2\",\"The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\",\"17419837\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02140\",\"afp4-1\",\"Decreased germination on high concentrations of glucose and sorbitol.\",\"18484180\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02230\",\"rgp1-1\",\"rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\",\"21478444\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02260\",\"tir3-1 RGLG1:rglg1 rglg2\",\"The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\",\"17586653\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02310\",\"sep2-1\",\"Non-described subtle phenotype.\",\"10821278\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02680\",\"atnbs1-1\",\"Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\",\"17672843\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02850\",\"CS3816\",\"The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\",\"17568770\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G02870\",\"vtc4-1\",\"ascorbate deficient\",\"16595667\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT3G03260\",\"hdg8-1\",\"No visible phenotype.\",\"16778018\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G14790\",\"pdd17\",\"Defective pollen development.\",\"19237690\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G15210\",\"bmy1-2\",\"Plants cold-shocked for 6h have an increased starch content compared to wildtype.\",\"16297066\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G15560\",\"cla1-1\",\"Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\",\"10982425\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G15570\",\"maa3\",\"Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\",\"18772186\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G15802\",\"Athspb-2\",\"Early flowering, reduced fertility, aborted seeds.\",\"20388662\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G15880\",\"esd4-2\",\"Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\",\"17513499\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G16420\",\"prz1-1\",\"Altered response to auxin and cytokinin\",\"12747832\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT4G16480\",\"atint4-2\",\"No visible phenotype.\",\"16603666\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT5G10480\",\"pas2-3\",\"Segregates 25% embryo lethal.\",\"18799749\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT5G10510\",\"plt3-1\",\"Short roots and shortened root meristem.\",\"17960244\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT5G11110\",\"kns2\",\"Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\",\"18779216\")\n",
    "%sql INSERT INTO germplasm_table (Locus, germplasm, phenotype, pumbed) VALUES (\"AT5G11260\",\"hy5-101\",\"Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\",\"16891401\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT1G01040\",\"DCL1\",\"332\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT1G01060\",\"LHY\",\"290\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT1G01140\",\"CIPK9\",\"223\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT1G01220\",\"FKGP\",\"190\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT2G03720\",\"MRH6\",\"189\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT2G03800\",\"GEK1\",\"196\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT2G04240\",\"XERICO\",\"256\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT2G05210\",\"POT1A\",\"221\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02130\",\"RPK2\",\"284\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02140\",\"TMAC2\",\"300\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02230\",\"RGP1\",\"301\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02260\",\"BIG\",\"279\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02310\",\"SEP2\",\"175\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02680\",\"NBS1\",\"190\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02850\",\"SKOR\",\"234\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G02870\",\"VTC4\",\"311\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT3G03260\",\"HDG8\",\"194\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G14790\",\"SUV3\",\"312\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G15210\",\"BAM5\",\"313\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G15560\",\"DXS\",\"219\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G15570\",\"MAA3\",\"294\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G15802\",\"HSBP\",\"254\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G15880\",\"ESD4\",\"265\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G16420\",\"ADA2B\",\"279\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT4G16480\",\"INT4\",\"284\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT5G10480\",\"PAS2\",\"301\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT5G10510\",\"AIL6\",\"310\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT5G11110\",\"SPS2\",\"232\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT5G11260\",\"HY5\",\"221\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT5G11510\",\"MYB3R-4\",\"336\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT5G12200\",\"PYD2\",\"310\")\n",
    "%sql INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES (\"AT5G13290\",\"CRN\",\"189\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_Exam_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>LocusGene_table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm_table</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('LocusGene_table',), ('germplasm_table',)]"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show tables\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 4:\n",
    "\n",
    "\n",
    "    1.Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "    2.Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "    3.Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "    4.Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule!\n",
    "\n",
    "All reports should be written to the same file. You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "57 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "59 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "(pymysql.err.ProgrammingError) (1064, \"You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'AS Number_of_Matches FROM LocusGene_table,germplasm_table WHERE LocusGene_table.' at line 1\")\n",
      "[SQL: SELECT COUNT * AS Number_of_Matches FROM LocusGene_table,germplasm_table WHERE LocusGene_table.Locus=germplasm_table.locus]\n",
      "(Background on this error at: http://sqlalche.me/e/f405)\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>AVG (LocusGene_table.ProteinLength)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>257.9649</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(Decimal('257.9649'),)]"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#1\n",
    "%sql SELECT * FROM LocusGene_table, germplasm_table WHERE LocusGene_table.Locus=germplasm_table.locus\n",
    "#2\n",
    "%sql SELECT * FROM LocusGene_table,germplasm_table WHERE LocusGene_table.Locus=germplasm_table.locus AND LocusGene_table.Gene=\"SKOR\" OR LocusGene_table.Gene=\"MAA3\"\n",
    "#3\n",
    "%sql SELECT COUNT * AS Number_of_Matches FROM LocusGene_table,germplasm_table WHERE LocusGene_table.Locus=germplasm_table.locus\n",
    "#4\n",
    "%sql SELECT AVG (LocusGene_table.ProteinLength) FROM LocusGene_table,germplasm_table WHERE LocusGene_table.Locus=germplasm_table.locus \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
