{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql #As we are using sqlMagic, we need to execute the connection lines.\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 1: Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Germplasm.tsv is based on AGI Locus Code\n",
      "LocusGene.tsv is based on AGI Locus Code\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "text= open(\"Germplasm.tsv\", \"r\")\n",
    "text1=open(\"LocusGene.tsv\",\"r\")\n",
    "textreader=text.read()\n",
    "text1reader=text1.read()\n",
    "\n",
    "x=\"AT.G.+\\t\"\n",
    "if re.search(x, textreader):\n",
    "    print(\"Germplasm.tsv is based on AGI Locus Code\")\n",
    "else:\n",
    "    print (\"Germplasm.tsv is not based on AGI Locus Code!!\")\n",
    "\n",
    "if re.search(x, text1reader):\n",
    "    print(\"LocusGene.tsv is based on AGI Locus Code\")\n",
    "else:\n",
    "    print (\"Germplasm.tsv is not based on AGI Locus Code!!\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 2: Design and create the database.¶\n",
    "\n",
    "    It should have two tables - one for each of the two data files.\n",
    "    The two tables should be linked in a 1:1 relationship\n",
    "    you may use either sqlMagic or pymysql to build the database\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "(pymysql.err.ProgrammingError) (1007, \"Can't create database 'Exam_2_modified'; database exists\")\n",
      "[SQL: create database Exam_2_modified;]\n",
      "(Background on this error at: http://sqlalche.me/e/f405)\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_Exam_2_modified</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>LocusGene_table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm_table</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('LocusGene_table',), ('germplasm_table',)]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases\n",
    "%sql create database Exam_2_modified;\n",
    "%sql show databases\n",
    "%sql use Exam_2_modified;\n",
    "%sql show tables;\n",
    "%sql CREATE TABLE germplasm_table(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, germplasm VARCHAR(100) NOT NULL, phenotype VARCHAR(1000) NOT NULL, pumbed INTEGER NOT NULL, LocusGene_id INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm_table \n",
    "%sql show tables\n",
    "%sql CREATE TABLE LocusGene_table (id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(10) NOT NULL, Gene VARCHAR(10) NOT NULL, ProteinLength INTEGER NOT NULL);\n",
    "%sql describe LocusGene_table\n",
    "%sql show tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 3: Fill the database¶\n",
    "\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block (<ipython-input-95-3659d0c1051c>, line 10)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-95-3659d0c1051c>\"\u001b[0;36m, line \u001b[0;32m10\u001b[0m\n\u001b[0;31m    sql=\"\"\"INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES ('{}','{}','{}')\"\"\".format(row [\"Locus\"],row[\"Gene\"],row [\"ProteinLength\"])\u001b[0m\n\u001b[0m      ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m expected an indented block\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "connection=pymysql.connect(host='localhost', user='root', password='root', db='Exam_2_modified', charset='utf8mb4', cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit(true)\n",
    "try:\n",
    "    with open('LocusGene.tsv') as csvfile:\n",
    "        LocusGene1 = csv.DictReader(csvfile, fieldnames=(\"Locus\", \"Gene\",\"ProteinLength\"))\n",
    "        print(type(LocusGene1))\n",
    "        for row in LocusGene1:\n",
    "        sql=\"\"\"INSERT INTO LocusGene_table (Locus, Gene, ProteinLength) VALUES ('{}','{}','{}')\"\"\".format(row [\"Locus\"],row[\"Gene\"],row [\"ProteinLength\"])\n",
    "        cursor.execute(sql)\n",
    "    with open ('Germplasm.tsv') as csvfile:\n",
    "        Germplasm1=csv.DictReader(csvfile,fieldnames=(\"Locus\",\"germplasm\",\"phenotype\",\"pubmed\"))\n",
    "        print(type(Germplasm1))\n",
    "        for row in Germplasm1:\n",
    "        sql=\"\"\"INSERT INTO germplasm_table (Locus,germplasm, phenotype, pumbed) VALUES ('{}','{}','{}','{}')\"\"\".format(row[\"Locus\"],row[\"germplasm\"], row[\"phenotype\"],row[\"pubmed\"])\n",
    "        cursor.execute(sql)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem 4: Create reports, written to a file\n",
    "\n",
    "    Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "    Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "    Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "    Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block (<ipython-input-111-623f01c6b5a1>, line 11)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-111-623f01c6b5a1>\"\u001b[0;36m, line \u001b[0;32m11\u001b[0m\n\u001b[0;31m    sql=\"\"\"SELECT (LocusGene_table.Locus,LocusGene_table.Gene, LocusGene_table.ProteinLength, germplasm_table.germplasm, germplasm_table.phenotype, germplasm_table.pubmed) FROM LocusGene_table, germplasm_table WHERE LocusGene_table.Locus=germplasm_table.locus\"\"\"\u001b[0m\n\u001b[0m      ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m expected an indented block\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import csv\n",
    "import io\n",
    "connection=pymysql.connect(host='localhost', user='root', password='root', db='Exam_2_modified', charset='utf8mb4', cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "results= ('results.csv', 'w', newline='')\n",
    "results.writerow([\"Locus\", \"Gene\", \"ProteinLength\", \"germplasm\",\"phenotype\",\"pubmed\"])  # write a header!\n",
    "p\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        print(results)\n",
    "        sql=\"SELECT (LocusGene_table.Locus,LocusGene_table.Gene, LocusGene_table.ProteinLength, germplasm_table.germplasm, germplasm_table.phenotype, germplasm_table.pubmed) FROM LocusGene_table, germplasm_table WHERE LocusGene_table.Locus=germplasm_table.Locus\"\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "    csvfile = io.open(\"results.csv\")\n",
    "    print(csvfile.read())\n",
    "    csvfile.close()\n",
    "try:\n",
    "    with open('results.csv', 'a', newline='') as csvfile:\n",
    "        results1=(csvfile)\n",
    "        for line in results1:\n",
    "        sql=\"\"\"SELECT * FROM LocusGene_table,germplasm_table WHERE LocusGene_table.Locus=germplasm_table.locus AND LocusGene_table.Gene=(\"SKOR\") OR LocusGene_table.Gene=(\"MAA3\")\"\"\"\n",
    "        cursor.execute(sql)\n",
    "    csvfile = io.open(\"results.csv\")\n",
    "    print(csvfile.read())\n",
    "    csvfile.close()\n",
    "try:\n",
    "    with open('results.csv', 'a', newline='') as csvfile:\n",
    "        results2=(csvfile)\n",
    "        for line in results2:\n",
    "        sql=\"\"\"SELECT COUNT * AS Number_of_Matches FROM LocusGene_table,germplasm_table WHERE LocusGene_table.Locus=germplasm_table.Locus\"\n",
    "        cursor.execute(sql)\n",
    "    csvfile = io.open(\"results.csv\")\n",
    "    print(csvfile.read())\n",
    "    csvfile.close()\n",
    "try:\n",
    "    with open('results.csv', 'a', newline='') as csvfile:\n",
    "        results3=(csvfile)\n",
    "        for line in results3:\n",
    "        sql=\"\"\"SELECT AVG (LocusGene_table.ProteinLength) FROM LocusGene_table,germplasm_table WHERE LocusGene_table.Locus=germplasm_table.Locus \"\"\"\n",
    "        cursor.execute(sql)\n",
    "    csvfile = io.open(\"results.csv\")\n",
    "    print(csvfile.read())\n",
    "    csvfile.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
